File: 8_kyu/Cat_years_Dog_years.c
// Duration in
// - human year
// - cat years
// - dog years
typedef struct Human_Cat_Dog_Years
{
  unsigned human_years, cat_years, dog_years;
} years;

// Converts y by the constants c.
static unsigned convert ( const unsigned y, const unsigned c[3] )
{
  switch( y ) {
      case 0 :
        return 0;
      case 1 :
        return c[0];
      case 2 :
        return c[0] + c[1];
      default :
        return c[0] + c[1] + (y-2) * c[2];
  }
}

// Calculate cat years and dog years.
years human_years_cat_years_dog_years(const unsigned human)
{
  // Constants to convert years
  static const unsigned cat_c [3] = {15, 9, 4};
  static const unsigned dog_c [3] = {15, 9, 5};
  // Caclculate the years
  const years result = {
    .human_years = human,
    .cat_years = convert( human, cat_c ),
    .dog_years = convert( human, dog_c ) };
  return result;
}

File: 8_kyu/Cat_years_Dog_years.py
def human_years_cat_years_dog_years(human_years: float) -> list[float,float,float]:
    """
    Calculates cat year and dog years.
    human_year -- standard year count.
    return -- list of human, cat and dog years.
    """
    if human_years <= 1:
        return [human_years, 15*human_years, 15*human_years]
    elif human_years <= 2:
        return [human_years, 15+9*(human_years-1), 15+9*(human_years-1)]
    else:
        return [human_years, 24+4*(human_years-2), 24+5*(human_years-2)]
